### PR TITLE
New leader change tests

### PR DESCRIPTION
**Pull request by ChrisKujawa**
## Description

Add new parameterized test where as before step either a message is published or process is deployed (or other things), the leader is stopped and in the after step the process instance is completed on the new leader.

 Tests are:

 * message correlation after leader change
 * timer trigger after leader change
 * job completion after leader change
 * instance creation after leader change

Additionally there is a test with including snapshot and one with several restarts.

## Related issues

related to https://github.com/camunda-cloud/zeebe/issues/7696

## Definition of Done

_Not all items need to be done depending on the issue and the pull request._

Code changes:
* [ ] The changes are backwards compatibility with previous versions
* [ ] If it fixes a bug then PRs are created to [backport](https://github.com/zeebe-io/zeebe/compare/stable/0.24...develop?expand=1&template=backport_template.md&title=[Backport%200.24]) the fix to the last two minor versions. You can trigger a backport by assigning labels (e.g. `backport stable/0.25`) to the PR, in case that fails you need to create backports manually.

Testing:
* [ ] There are unit/integration tests that verify all acceptance criterias of the issue
* [ ] New tests are written to ensure backwards compatibility with further versions
* [ ] The behavior is tested manually
* [ ] The change has been verified by a QA run
* [ ] The impact of the changes is verified by a benchmark 

Documentation: 
* [ ] The documentation is updated (e.g. BPMN reference, configuration, examples, get-started guides, etc.)
* [ ] New content is added to the [release announcement](https://drive.google.com/drive/u/0/folders/1DTIeswnEEq-NggJ25rm2BsDjcCQpDape)

### CODE REVIEW

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+import org.junit.Test;
+import org.junit.rules.RuleChain;
+import org.junit.rules.Timeout;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class CompleteProcessInstanceAfterLeaderChangeTest {
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:31]: :question: Do we need to test this exhaustively on different types of processes? Are they not tested anywhere else? It feels like there must be more focused unit test for these things. No blockers, btw.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  public String name;
+
+  @Parameter(1)
+  public Consumer<GrpcClientRule> beforeRestart;
+
+  @Parameter(2)
+  public BiConsumer<ClusteringRule, GrpcClientRule> afterRestart;
+
+  @Parameters(name = "{index}: {0}")
+  public static Object[][] restartAction() {
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:52]:
```suggestion
  public static Object[][] actions() {
```

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            (clusteringRule, clientRule) -> {
+              final var processDefinitionKey =
+                  clientRule.deployProcess(
+                      Bpmn.createExecutableProcess("process")
+                          .startEvent()
+                          .intermediateCatchEvent()
+                          .message(
+                              msg -> {
+                                msg.name("msg").zeebeCorrelationKey("=key");
+                              })
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:91]:
```suggestion
                          .message(msg -> msg.name("msg").zeebeCorrelationKey("=key"))
```

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+                      Bpmn.createExecutableProcess("process")
+                          .startEvent()
+                          .intermediateCatchEvent()
+                          .timerWithDuration("PT15S")
+                          .endEvent()
+                          .done());
+              clientRule.createProcessInstance(processDefinitionKey);
+            },
+        (BiConsumer<ClusteringRule, GrpcClientRule>)
+            (clusteringRule, clientRule) -> {
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:119]: can be replaced with expression lambda

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+                        clusteringRule.getClock().addTime(Duration.ofSeconds(15));
+                        ZeebeAssertHelper.assertProcessInstanceCompleted("process");
+                      });
+            }
+      },
+      new Object[] {
+        "complete job after restart",
+        (Consumer<GrpcClientRule>) (clientRule) -> clientRule.createSingleJob("testTask"),
+        (BiConsumer<ClusteringRule, GrpcClientRule>)
+            (clusteringRule, clientRule) -> {
```
Comment (deepthidevaki) [inline qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:132]: can be replaced with expression lambda

**Review by deepthidevaki — APPROVED** [review]
Thanks. Just have some minor comments.

File: qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java (hunk 1 above)
Comment (Zelldon) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:31]: The point for me is that these things are all pointing to different things on the system, like message correlation after leader change or trigger timers, where the timer need to be scheduled etc. I would feel better if we had a test like that.

Yes we have something on the engine level, but this doesn't respect other parts.

Comment (Zelldon) [reply qa/integration-tests/src/test/java/io/camunda/zeebe/it/clustering/CompleteProcessInstanceAfterLeaderChangeTest.java:31]: Unfortunately the IT test take longer to execute

**Pull request merged**